### DISCUSSION TITLE
Non-deno-deploy deployments

### DISCUSSION THREAD

**Discussion by Eyal-Shalev** [freshframework/fresh#200] [Ideas]
Is it possible (or planned) to deploy a fresh application to a custom server that isn't part of the deno deploy framework?

**Comment by lucacasonato** [comment]
Yup, you can run fresh on any server using `deno run`. The only thing you need to do to enable "production mode" is to set a `DENO_DEPLOYMENT_ID` environment variable with a unique hash of the project files (for example a git commit hash).

**Comment by Utwo** [comment]
@lucacasonato I can make a `Dockerfile` that can server as an example of non-deno deploy if you want to add it to the docs.

**Reply by lucacasonato** [reply]
Feel free to open a PR.

**Comment by Utwo** [comment]
Alright, I think we can use this as a starting point.
```Dockerfile
FROM denoland/deno:alpine-1.23.2

ARG denoDeploymentId=unique_id
ENV DENO_DEPLOYMENT_ID=$denoDeploymentId

EXPOSE 8000
WORKDIR /app
USER deno

COPY --chown=deno . .
RUN deno cache main.ts
RUN echo ${DENO_DEPLOYMENT_ID}
CMD ["run", "--allow-net","--allow-read","--allow-env","--allow-write","--allow-run","main.ts"]
```

To build, we can run:
```bash
$ docker build . -t fresh --build-arg denoDeploymentId=123
```

Do you want to add this to the docs page? https://fresh.deno.dev/docs/getting-started/deploy-to-production

**Comment by PierBover** [comment]
I'm a Deno noob trying to deploy a Fresh project to Fly with Docker.

This is from the official [Deno image](https://hub.docker.com/r/denoland/deno):

```
FROM denoland/deno:1.10.3

# The port that your application listens to.
EXPOSE 1993

WORKDIR /app

# Prefer not to run as root.
USER deno

# Cache the dependencies as a layer (the following two steps are re-run only when deps.ts is modified).
# Ideally cache deps.ts will download and compile _all_ external files used in main.ts.
COPY deps.ts .
RUN deno cache deps.ts

# These steps will be re-run upon each file change in your working directory:
ADD . .
# Compile the main app so that it doesn't need to be compiled each startup/entry.
RUN deno cache main.ts

CMD ["run", "--allow-net", "main.ts"]
```

But my Fresh project doesn't have any `deps.ts` and of course I get this error:

```
 => ERROR [2/6] COPY deps.ts .    
```

Should I just ignore that line or copy `deno.json`?

Edit:

Tried this but the build is failing too.

```
COPY deno.json .
RUN deno cache deno.json
```

```
error: An unsupported media type was attempted to be imported as a module.   
```

So I'm guessing the issue is that the official Docker image doesn't have the latest Deno version? Weird because apparently the Docker image was last updated 15 hours ago.

**Reply by marvinhagemeister** [reply]
The `deno cache` command expects your entrypoint file to be an argument, not `deno.json`. Something like `deno cache main.ts` for Fresh. It's basically similar to `npm install` with the difference, that it only downloads the dependencies that it can reach from the entrypoint file that was passed to it.

I'd assume that the `deps.ts` step is not necessary for Fresh.

**Reply by PierBover** [reply]
Thanks Marvin.

I did try removing those `deps.ts` lines and simply doing `deno cache main.ts` but then got this error:

```
#7 0.596 Check file:///main.ts
#7 2.296 error: TS2305 [ERROR]: Module '"deno:///missing_dependency.d.ts"' has no exported member 'start'.
#7 2.296 import { start } from "$fresh/server.ts";
#7 2.296          ~~~~~
#7 2.296     at file:///main.ts:9:10
```

I have no idea what's going on. I guess everyone must be using Deno with Deno Deploy because I haven't been able to find much information on how to solve this.

Would be great if the Fresh docs had a section on how to deploy with Docker.
